gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagActionReminderScheduler.java

```diff
@@ -57,20 +64,95 @@ public DagActionReminderScheduler(StdSchedulerFactory schedulerFactory, Optional
+          jobDataMap.getString(FLOW_ACTION_TYPE_KEY));
+
+      log.info("DagProc reminder triggered for (flowGroup: " + flowGroup + ", flowName: " + flowName
+          + ", flowExecutionId: " + flowId + ", jobName: " + jobName +")");
+
+      DagActionStore.DagAction dagAction = new DagActionStore.DagAction(flowGroup, flowName, flowId, jobName,
+          dagActionType);
+
+      try {
+        DagManagement dagManagement = GobblinServiceManager.getClass(DagManagement.class);
+        dagManagement.addDagAction(dagAction);
+      } catch (IOException e) {
+        log.error("Failed to add DagAction to DagManagement. Action: {}", dagAction);
+      }
+    }
+  }
+
+  /**
+   * Creates a key for the reminder job by concatenating all dagAction fields
+   */
+  public static String createDagActionReminderKey(DagActionStore.DagAction dagAction) {
+    return createDagActionReminderKey(dagAction.getFlowName(), dagAction.getFlowGroup(), dagAction.getFlowExecutionId(),
+        dagAction.getJobName(), dagAction.getDagActionType());
+  }
+
+  /**
+   * Creates a key for the reminder job by concatenating flowName, flowGroup, flowExecutionId, jobName, dagActionType
+   * in that order
+   */
+  public static String createDagActionReminderKey(String flowName, String flowGroup, String flowId, String jobName,
+      DagActionStore.DagActionType dagActionType) {
+    return String.format("%s.%s.%s.%s.%s", flowGroup, flowName, flowId, jobName, dagActionType);
+  }
+
+  /**
+   * Creates a jobDetail containing flow and job identifying information in the jobDataMap, uniquely identified
+   *  by a key comprised of the dagAction's fields.
+   */
+  public static JobDetail createReminderJobDetail(DagActionStore.DagAction dagAction) {
+    JobDataMap dataMap = new JobDataMap();
+    dataMap.put(ConfigurationKeys.FLOW_NAME_KEY, dagAction.getFlowName());
+    dataMap.put(ConfigurationKeys.FLOW_GROUP_KEY, dagAction.getFlowGroup());
+    dataMap.put(ConfigurationKeys.JOB_NAME_KEY, dagAction.getJobName());
+    dataMap.put(ConfigurationKeys.FLOW_EXECUTION_ID_KEY, dagAction.getFlowExecutionId());
+    dataMap.put(ReminderJob.FLOW_ACTION_TYPE_KEY, dagAction.getDagActionType());
+
+    return JobBuilder.newJob(ReminderJob.class)
+        .withIdentity(createDagActionReminderKey(dagAction), dagAction.getFlowName())
+        .usingJobData(dataMap)
+        .build();
+  }
+
+  /**
+   * Creates a Trigger object with the same key as the ReminderJob (since only one trigger is expected to be associated
+   * with a job at any given time) that should fire after `reminderDurationMillis` millis.
+   */
+  public static Trigger createReminderJobTrigger(DagActionStore.DagAction dagAction, long reminderDurationMillis) {
+    Trigger trigger = TriggerBuilder.newTrigger()
+        .withIdentity(createDagActionReminderKey(dagAction), dagAction.getFlowName())
+        .startAt(new Date(System.currentTimeMillis() + reminderDurationMillis))
```

Comment:
hard-coded "clocks" make testing challenging.  AFAICT, this method here should just take the desired target time in millis (rather than the offset).

that leaves the caller, `scheduleReminder`, the need for obtaining the current time.  for that, I suggest a `Supplier<Long> getCurrentTimeMillis` member.  pair that with a ctor form taking such a param (when testing).  the ctor w/o that extra param would simply pass in `System::currentTimeMillis`